cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,183 @@
-        - mul               (len(Operator.args) == 2)
+        - NumVar
         - sum
         - wsum
 
+Indicator constraints:
+----------------------
+- BoolVar -> LinExpr == Constant
+- BoolVar -> LinExpr >= Constant
+- BoolVar -> LinExpr <= Constant
 
+- BoolVar -> GenExpr                    (GenExpr.name in supported, GenExpr.is_bool())
+- BoolVar -> GenExpr >= Var/Constant    (GenExpr.name in supported)
+- BoolVar -> GenExpr <= Var/Constant    (GenExpr.name in supported)
+- BoolVar -> GenExpr == Var/Constant    (GenExpr.name in supported)
 
-Indicator constraints
---------------------------
-- Boolvar -> LinExpr == Constant
-- Boolvar -> LinExpr >= Constant
-- Boolvar -> LinExpr <= Constant
+Where BoolVar is a boolean variable or its negation.
 
+General comparisons or expressions
+-----------------------------------
+- GenExpr                               (GenExpr.name in supported, GenExpr.is_bool())
+- GenExpr == Var/Constant               (GenExpr.name in supported)
+- GenExpr <= Var/Constant               (GenExpr.name in supported)
+- GenExpr >= Var/Constant               (GenExpr.name in supported)
 
-General Constraints
----------------------------
-GenExpr == var
-
-    GenExpr can be any of: min, max, pow, abs
 
 """
+import copy
 import numpy as np
 
-from .reification import only_bv_implies
-from .flatten_model import flatten_constraint, get_or_make_var, negated_normal
+from .flatten_model import flatten_constraint, get_or_make_var
 
 from ..expressions.core import Comparison, Operator, _wsum_should, _wsum_make
 from ..expressions.globalconstraints import GlobalConstraint
-from ..expressions.utils import is_any_list, is_num
+from ..expressions.utils import is_any_list, is_num, eval_comparison, is_bool
+
 from ..expressions.variables import _BoolVarImpl, boolvar, NegBoolView, _NumVarImpl
 
-def linearize_constraint(cpm_expr):
+def linearize_constraint(cpm_expr, supported={"sum","wsum"}):
     """
     Transforms all constraints to a linear form.
     This function assumes all constraints are in 'flat normal form'.
     Only apply after 'cpmpy.transformations.flatten_model.flatten_constraint()'.
     """
 
     if is_any_list(cpm_expr):
-        lin_cons = [linearize_constraint(expr) for expr in cpm_expr]
+        lin_cons = [linearize_constraint(expr, supported=supported) for expr in cpm_expr]
         return [c for l in lin_cons for c in l]
 
+    # boolvar
     if isinstance(cpm_expr, _BoolVarImpl):
         if isinstance(cpm_expr, NegBoolView):
-            return [cpm_expr._bv <= 0]
-        return [cpm_expr >= 1]
+            return [sum([cpm_expr._bv]) <= 0]
+        return [sum([cpm_expr]) >= 1]
 
+    # conjunction
     if cpm_expr.name == "and":
-        if all(arg.is_bool() for arg in cpm_expr.args):
-            return [sum(cpm_expr.args) >= len(cpm_expr.args)]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+        return [sum(cpm_expr.args) >= len(cpm_expr.args)]
```

Comment:
Ideally, no negated variables are left in these sums. E.g., the negation is removed for `# boolvar`. Here it is not. I'd vote to remove it at this point (it is typically only a matter of changing signs and rhs), as it would fit more closely with a true linearization.

Reply:
I see that below, the `only_positive_bv` takes care of this. Good. But then we do not need to do it above either.